compression_verifier.go

```diff
@@ -0,0 +1,238 @@
+			if column.Name == pkColumn {
+				continue
+			}
+			// Check if column is configured as compressed and decompress if necessary
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				decompressedColData, err := c.Decompress(table, column.Name, algorithm, rowData[idx].([]byte))
+				if err != nil {
+					return nil, err
+				}
+				decompressedRowData[uint64(idx)] = decompressedColData
+			} else {
+				// Do not decompress the column, add it to the decompressedRowData to be fingerprinted
+				decompressedRowData[uint64(idx)] = rowData[idx].([]byte)
+			}
+		}
+
+		// Hash the data of the row to be added to the result set
+		decompressedRowHash, err := c.HashRow(decompressedRowData)
+		if err != nil {
+			return nil, err
+		}
+
+		resultSet[pk] = decompressedRowHash
+	}
+
+	return resultSet, nil
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
+	quotedPK := quoteField(pkColumn)
+	sql, args, err := rowSelector(columns, pkColumn).
+		From(QuotedTableNameFromString(schema, table)).
+		Where(sq.Eq{quotedPK: pks}).
+		OrderBy(quotedPK).
+		ToSql()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// This query must be a prepared query. If it is not, querying will use
+	// MySQL's plain text interface, which will scan all values into []uint8
+	// if we give it []interface{}.
+	stmt, err := db.Prepare(sql)
+	if err != nil {
+		return nil, err
+	}
+
+	defer stmt.Close()
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+// Decode will apply the configured decompression algorithm to the configured columns data
+func (c *CompressionVerifier) Decompress(table, column, algorithm string, compressed []byte) ([]byte, error) {
+	var decompressed []byte
+	switch algorithm {
+	case CompressionSnappy:
+		return snappy.Decode(decompressed, compressed)
+	default:
+		return nil, UnsupportedCompressionError{
+			table:     table,
+			column:    column,
+			algorithm: algorithm,
+		}
+	}
+
+}
+
+// HashRow will fingerprint the non-primary columns of the row to verify data equality
+func (c *CompressionVerifier) HashRow(decompressedRowData map[uint64][]byte) ([]byte, error) {
+	if len(decompressedRowData) == 0 {
+		return nil, errors.New("Row data to fingerprint must not be empty")
+	}
+
+	hash := md5.New()
+	var rowFingerprint []byte
+	for _, colData := range decompressedRowData {
+		rowFingerprint = append(rowFingerprint, colData...)
+	}
+
+	hash.Write(rowFingerprint)
+	return []byte(hex.EncodeToString(hash.Sum(nil))), nil
+}
+
+func rowSelector(columns []schema.TableColumn, pkColumn string) sq.SelectBuilder {
+	quotedPK := quoteField(pkColumn)
+
+	columnStrs := make([]string, len(columns))
+	for idx, column := range columns {
+		if column.Name == pkColumn {
+			continue
+		}
+
+		quotedCol := normalizeAndQuoteColumn(column)
+		columnStrs[idx] = fmt.Sprintf("COALESCE(%s, 'NULL')", quotedCol)
```

Comment:
This line just writes `NULL` instead of having an empty value for the column. Turns out we can remove it and just use `NULL`s. Why is this needed for the `IterativeVerifier`?

Reply:
Basically because you can't `MD5(NULL)` in MySQL, if I remember correctly.

Reply:
Hmm that doesn't appear to be the case.

```mysql
mysql> select md5(null)\G
*************************** 1. row ***************************
md5(null): NULL
1 row in set (0.00 sec)

mysql>
```

Reply:
Hm then it was probably the `CONCAT` that needed this